### DISCUSSION TITLE
[Question] How to draw a straight waveguide PCell with polygon method?

### DISCUSSION THREAD

**Discussion by luomeng007** [gdsfactory/gdsfactory#3244] [Q&A] [answered]
I have tried four methods to draw a straight waveguide pcell, like below:   
1. gf.kdb.DPolygon()  
2. add_polygon()  
3. gf.components.rectangle()
4. gf.path.straight()

All methods above can successfully draw a straight waveguide pcell. However, if I export the structure to KLayout view. The figure is shown like below:
![image](https://github.com/user-attachments/assets/a4075fde-337c-4b28-a5ca-39e2d0f2641e)
See, only two points for constructing this straight waveguide. If I double left click on this pcell. The figure below appears, obviously, this is a box method to draw straight waveguide.
![image](https://github.com/user-attachments/assets/9d7d1344-ef86-47dc-8efa-999773458156)
So, my question is, even I used polygon method to draw straight waveguide in GDSFactory, but when export straight waveguide into KLayout, the straight waveguie is build by using box method in KLayout.  But I want it is build by polygon method, like figure below:
![image](https://github.com/user-attachments/assets/da9b9e4f-c132-436f-a14d-9f8d0a002769)

Anyone can help me to deal with this? 

Besides, can we add a question label in issue, I only find enhancement and bug labels there!

**Comment by sebastian-goeldi** [comment, marked as answer]
> Besides, can we add a question label in issue, I only find enhancement and bug labels there!

That's what discussions are for ;)

The reason you see a box is that this is an optimization from klayout. A box is internally just a more straight forward representation of a rectangle polygon (a box you can describe by two points instead of using 4 when using a polygon).

You can force klayout however to not do this optimization (not sure why you would ever want to do that, but here goes):

## pure dbu case (i.e. work in database units, normally this is `1 dbu == 1nm`, i.e. `dbu == 0.001`)

```python
import kfactory as kf

pts = [
    kf.kdb.Point(0, 0),
    kf.kdb.Point(5, 0),
    kf.kdb.Point(10, 0),
    kf.kdb.Point(15, 0),
    kf.kdb.Point(20, 0),
    kf.kdb.Point(25, 0),
    kf.kdb.Point(25, 10),
    kf.kdb.Point(20, 10),
    kf.kdb.Point(15, 10),
    kf.kdb.Point(10, 10),
    kf.kdb.Point(5, 10),
    kf.kdb.Point(0, 10),
]

d1 = kf.kdb.Polygon(pts)
print(f"{d1=}")

d2 = kf.kdb.Polygon(
    pts,
    raw=True,
)
print(f"{d2=}")

c = kf.KCell()
p = c.shapes(kf.kdb.LayerInfo(1, 0)).insert(d1)
print(p)
p = c.shapes(kf.kdb.LayerInfo(1, 0)).insert(d2)
print(p)
```

```
d1=(0,0;0,10;25,10;25,0)
d2=(0,0;0,10;5,10;10,10;15,10;20,10;25,10;25,0;20,0;15,0;10,0;5,0)
polygon (0,0;0,10;25,10;25,0)
polygon (0,0;0,10;5,10;10,10;15,10;20,10;25,10;25,0;20,0;15,0;10,0;5,0)
```

As you can see the optimization happens already  when creating the polygon, if we pass `raw=True`, it will tell KLayout to not optimize it.

## um based version

`raw=True` doesn't have an effect on um based DPolygons (I assume because it would be inefficient and inaccurate in many cases). But we can coerce the DPolygon to a Polygon first which doesn't optimize the polygon and keeps the points.

```python
import kfactory as kf

pts = [
    kf.kdb.DPoint(0, 0),
    kf.kdb.DPoint(5, 0),
    kf.kdb.DPoint(10, 0),
    kf.kdb.DPoint(15, 0),
    kf.kdb.DPoint(20, 0),
    kf.kdb.DPoint(25, 0),
    kf.kdb.DPoint(25, 10),
    kf.kdb.DPoint(20, 10),
    kf.kdb.DPoint(15, 10),
    kf.kdb.DPoint(10, 10),
    kf.kdb.DPoint(5, 10),
    kf.kdb.DPoint(0, 10),
]

d1 = kf.kdb.DPolygon(pts)
print(f"{d1=}")

c = kf.KCell()
p = c.shapes(kf.kdb.LayerInfo(1, 0)).insert(d1)
print(p)
p = c.shapes(kf.kdb.LayerInfo(1, 0)).insert(d1.to_itype(c.kcl.dbu))
print(p)
```

```
d1=(0,0;0,10;5,10;10,10;15,10;20,10;25,10;25,0;20,0;15,0;10,0;5,0)
polygon (0,0;0,10000;25000,10000;25000,0)
polygon (0,0;0,10000;5000,10000;10000,10000;15000,10000;20000,10000;25000,10000;25000,0;20000,0;15000,0;10000,0;5000,0)
```

EDIT: Apologies for using kfactory here. I forgot to change it back to gdsfactory code at the end. But everything `kf` can be replaced with `gf ` except for `c = kf.KCell()` there use `c = gf.Component()`

**Reply by luomeng007** [reply]
Thank you for your reply. I will have a try for this. 

Now, I get this, next time, when I have question, I will mark it as a discussion. :)

For reason I want to do this, I can share my idea with you. For example, if we have one straight waveguide and an u-arc structure, and we want to get the minimum distance between them, normally, we need to use a lot of points on this straight waveguide. If we only use four points or two points to descript this straight waveguide, we cannot use program to get the minimum distance  between them. You know, the minimum distance is needed for optimize the die size when we draw layout of chip.

Actually, I have already write my own program to get the minimum distance between two random devices in anohter software. Currently, I want to migrate my code to GDSFactory. 

I also hope you can add this function to GDSFactory, it will be a useful tool for user to optimize the die figure to get the limitation size.

Finally, I must admit that I really like this software, you guys are really awesome. Hope this software can become stronger as soon as possible.

**Reply by sebastian-goeldi** [reply]
It kind of is there already. Though it is a bit tricky. If you purely want point to point distance, it's easy. The overarching general case is more complex.

But essentially what you can do:

```python
# get the edge of the straight closest to the bend
# I am assuming a manhattan straight here, but it works with any one if you get the points similar to the bend
dbbox = my_straight.dbbox(relevant_layer)
dedge = gf.kdb.DEdge(gf.kdb.DPoint(dbbox.left,dbbox.top), gf.kdb.DPoint(dbbox.right, dbbox.top))
dpoint_region = gf.kdb.Region(my_bend.shapes(relevant_layer)).merge().hulls()
dpoints = [pt for polygon in dpoint_region.each() for pt in poly.to_itype(gf.kcl.dbu)]
min_dist_to_pts = min(dedge.distance_abs(pt) for pt in dpoints)
```

Note, this gives you the closes distance to the top edge of the straight. No need to have a non-optimized polygon for this.

**Reply by luomeng007** [reply]
> It kind of is there already. Though it is a bit tricky. If you purely want point to point distance, it's easy. The overarching general case is more complex.
> 
> But essentially what you can do:
> 
> ```python
> # get the edge of the straight closest to the bend
> # I am assuming a manhattan straight here, but it works with any one if you get the points similar to the bend
> dbbox = my_straight.dbbox(relevant_layer)
> dedge = gf.kdb.DEdge(gf.kdb.DPoint(dbbox.left,dbbox.top), gf.kdb.DPoint(dbbox.right, dbbox.top))
> dpoint_region = gf.kdb.Region(my_bend.shapes(relevant_layer)).merge().hulls()
> dpoints = [pt for polygon in dpoint_region.each() for pt in poly.to_itype(gf.kcl.dbu)]
> min_dist_to_pts = min(dedge.distance_abs(pt) for pt in dpoints)
> ```
> 
> Note, this gives you the closes distance to the top edge of the straight. No need to have a non-optimized polygon for this.

I tried to understant the code you show me. It does take me a lot of time. There are something that I cannot understant well.  
1.  `dbbox()` function, it should not have parameters, right? When I use `my_straight.dbbox()` I get result, but when I use `my_straight.dbbox(relevant_layer)`, It raise an error.
2.  for `my_bend.shapes(relevant_layer)`, this will give a `klayout.dbcore.Shapes` object. I must provide layer to `shapes()` function, or I would get an error. Howverer, when I created my_bend object, I have already provided layer information. Why I still need to provide layer information for `shapes()` function. I mean can we use `my_bend.shapes()` instead? just like `dbbox()` function.
3. `dpoints = [pt for polygon in dpoint_region.each() for pt in poly.to_itype(gf.kcl.dbu)]` this one, VS code raise some errors, I do not understant well what is `poly` here, seems there is no variable called `poly`, so I guess you mean `dpoints = [pt for polygon in dpoint_region.each() for pt in polygon.to_itype(gf.kcl.dbu)]`. However, after I run program, I get information as 'Polygon' object has no attribute 'to_itype'

Could you revise the code and give me some tips about this part?

**Reply by sebastian-goeldi** [reply]
1. `my_straight.dbbox(relevant_layer)` relevant_layer must be an index here not a tuple or anything, and that index must exist, i.e. `c.kcl.get_info(relevant_layer)` must return a layer info object.
2. shapes is a container for shapes, i.e. think of it similar to a region. shapes without an arg is not valid.
3. I made a typo in there, apologies. We want of course the um based DPolygon from a Polygon. `[pt for polygon in dpoint_region.each() for pt in polygon.to_dtype(gf.kcl.dbu)].each_point_hull()`. `DPolgyon.to_itype(dbu)` converts it to a Polygon and `Polygon.to_dtype(dbu)` converts a Polygon to a DPolygon

**Reply by luomeng007** [reply]
Thank you, I see now. For `c.kcl.get_info(relevant_layer)` part, as I tested, I find there are 44 generic defined layers as belowing:  

0. WAFER (999/0)  
1. WG (1/0)  
2. WGCLAD (111/0)  
3. SLAB150 (2/0)  
4. SHALLOW_ETCH (2/6)  
5. SLAB90 (3/0)  
6. DEEP_ETCH (3/6)  
7. DEEPTRENCH (4/0)  
8. GE (5/0)  
9. UNDERCUT (6/0)  
10. WGN (34/0)  
11. WGN_CLAD (36/0)  
12. N (20/0)  
13. NP (22/0)  
14. NPP (24/0)  
15. P (21/0)  
16. PP (23/0)  
17. PPP (25/0)  
18. GEN (26/0)  
19. GEP (27/0)  
20. HEATER (47/0)  
21. M1 (41/0)  
22. M2 (45/0)  
23. MTOP (49/0)  
24. VIAC (40/0)  
25. VIA1 (44/0)  
26. VIA2 (43/0)  
27. PADOPEN (46/0)  
28. DICING (100/0)  
29. NO_TILE_SI (71/0)  
30. PADDING (67/0)  
31. DEVREC (68/0)  
32. FLOORPLAN (64/0)  
33. TEXT (66/0)  
34. WG_PIN (1/10)  
35. PORTE (1/11)  
36. PORTH (70/0)  
37. SHOW_PORTS (1/12)  
38. LABEL_INSTANCE (206/0)  
39. LABEL_SETTINGS (202/0)  
40. TE (203/0)  
41. TM (204/0)  
42. DRC_MARKER (205/0)  
43. SOURCE (110/0)  
44. MONITOR (101/0) 

If user add their own layers, the layers will be put on 45, 46 and so on. However, I did not see any explanation about the order of these layers, maybe there should be a document or something?  

For the minimum distance part, I still need some time to try it according to your instructions. See you later. :)

**Reply by sebastian-goeldi** [reply]
There is no documentation on it because it's neither guaranteed to be in this order nor is it guaranteed to be this amount, you cannot infer that your layer will always be 45 and so on. If you are using the LayerEnum, it keeps track of the index for you, otherwise you need to get the index with `gf.kcl.layer(...)` where you pass the info object or the individual info parts (layer number + datatype / name) or the other way around where you get the info from the index with `get_info`.

The layers themselves come from the generic technology of gdsfactory afaik. The backend itself doesn't define any layers.

But if you are relying on the index being a certain indexed way is like inviting guests to a party and requiring them to show up in a certain order. It will almost never work ;)

(Essentially you can think of the backend like a small database. You also would never make an assumption about the serial primary key of a column. The equivalent would be to require 45 to always be the same object no matter which database you are querying for a layer, so you need a mapping dict/list such as `SELECT id, layer, datatype FROM layers;` which is the equivalent to the LayerEnum object)

**Reply by luomeng007** [reply]
Ok, I see, thx.   

Recently, I am quite busy one my job, I still have no time to explore the distance part. After I test that, I will come back.

**Reply by luomeng007** [reply]
> 3. I made a typo in there, apologies. We want of course the um based DPolygon from a Polygon. `[pt for polygon in dpoint_region.each() for pt in polygon.to_dtype(gf.kcl.dbu)].each_point_hull()`. `DPolgyon.to_itype(dbu)` converts it to a Polygon and `Polygon.to_dtype(dbu)` converts a Polygon to a DPolygon

I tested the code above, when I run the code,  some errors happened. If I understand correctly, the code `pt for polygon in dpoint_region.each()`, `pt` will get the points of the region right? But this seems not work, the result is still a whole area. Like:
`(0,-250;0,250;251,251;503,253;754,256;1005,260;1256,266;1507,273;1759,281;2261,301;2512,313;2763,327;3013,341;3264,357;3515,374;3766,393;4016,412;4266,433;4517,455;4767,479;5017,504;5267,530;5517,557;5766,585;6016,615;6265,646;6514,678;6763,712;7012,747;7261,783;7509,820;7758,859;8006,898;8254,939;8501,982;8749,1025;8996,1070;9243,1116;9737,1212;9983,1262;10229,1313;10475,1365;10965,1473;11210,1529;11455,1587;11699,1645;11943,1705;12187,1766;12431,1828;12674,1891;12917,1956;13159,2022;13401,2089;13643,2157;13885,2227;14367,2369;14847,2517;15087,2593;15326,2669;15565,2747;16041,2907;16279,2989;16516,3072;16753,3156;16989,3241;17225,3327;17461,3415;17696,3504;17930,3594;18165,3685;18398,3777;18632,3871;18864,3965;19097,4061;19328,4158;19560,4256;19790,4356;20021,4456;20250,4558;20480,4661;20708,4765;20937,4870;21164,4976;21618,5192;21844,5302;22070,5413;22294,5525;22519,5638;22743,5753;22966,5868;23410,6102;23632,6221;23853,6341;24073,6462;24292,6584;24511,6707;24730,6832;24948,6957;25165,7084;25381,7211;25597,7340;25812,7470;26026,7601;26240,7733;26453,7866;26666,8000;26878,8135;27089,8272;27299,8409;27509,8547;27718,8687;27926,8827;28134,8969;28341,9112;28547,9255;28752,9400;28957,9546;29161,9692;29364,9840;29567,9989;29768,10139;29969,10290;30369,10594;30568,10748;30765,10903;30962,11059;31159,11216;31354,11374;31549,11533;31743,11693;31936,11853;32128,12015;32320,12178;32700,12506;32889,12672;33077,12839;33264,13006;33451,13175;33636,13344;33821,13514;34005,13686;34188,13858;34370,14031;34551,14205;34732,14380;34911,14556;35090,14733;35267,14910;35444,15089;35620,15268;35795,15449;35969,15630;36142,15812;36314,15995;36486,16179;36656,16364;36825,16549;36994,16736;37161,16923;37328,17111;37494,17300;37822,17680;37985,17872;38147,18064;38307,18257;38467,18451;38626,18646;38784,18841;38941,19038;39097,19235;39252,19432;39406,19631;39710,20031;39861,20232;40011,20433;40160,20636;40308,20839;40454,21043;40600,21248;40745,21453;40888,21659;41031,21866;41173,22074;41313,22282;41453,22491;41591,22701;41728,22911;41865,23122;42000,23334;42134,23547;42267,23760;42399,23974;42530,24188;42660,24403;42789,24619;42916,24835;43043,25052;43168,25270;43293,25489;43416,25708;43538,25927;43659,26147;43779,26368;43898,26590;44132,27034;44247,27257;44362,27481;44475,27706;44587,27930;44698,28156;44808,28382;45024,28836;45130,29063;45235,29292;45339,29520;45442,29750;45544,29979;45644,30210;45744,30440;45842,30672;45939,30903;46035,31136;46129,31368;46223,31602;46315,31835;46406,32070;46496,32304;46585,32539;46673,32775;46759,33011;46844,33247;46928,33484;47011,33721;47093,33959;47253,34435;47331,34674;47407,34913;47483,35153;47631,35633;47773,36115;47843,36357;47911,36599;47978,36841;48044,37083;48109,37326;48172,37569;48234,37813;48295,38057;48355,38301;48413,38545;48471,38790;48527,39035;48635,39525;48687,39771;48738,40017;48788,40263;48884,40757;48930,41004;48975,41251;49018,41499;49061,41746;49102,41994;49141,42242;49180,42491;49217,42739;49253,42988;49288,43237;49322,43486;49354,43735;49385,43984;49415,44234;49443,44483;49470,44733;49496,44983;49521,45233;49545,45483;49567,45734;49588,45984;49607,46234;49626,46485;49643,46736;49659,46987;49673,47237;49687,47488;49699,47739;49719,48241;49727,48493;49734,48744;49740,48995;49744,49246;49747,49497;49749,49749;49750,50000;50250,50000;50249,49746;50247,49492;50244,49239;50240,48985;50234,48731;50227,48477;50219,48224;50209,47970;50198,47717;50186,47463;50172,47210;50158,46956;50142,46703;50106,46197;50086,45944;50065,45691;50019,45185;49994,44933;49968,44680;49940,44428;49911,44176;49881,43924;49850,43672;49817,43420;49783,43169;49748,42917;49712,42666;49674,42415;49635,42164;49595,41914;49554,41663;49511,41413;49467,41163;49422,40913;49375,40664;49327,40415;49278,40166;49228,39917;49176,39668;49124,39420;49070,39172;49014,38924;48958,38677;48900,38430;48841,38183;48780,37936;48719,37690;48656,37444;48592,37199;48527,36953;48460,36709;48393,36464;48324,36220;48182,35732;48109,35489;48035,35246;47960,35004;47884,34762;47806,34520;47727,34279;47647,34038;47566,33798;47484,33558;47400,33318;47315,33079;47229,32840;47142,32602;46964,32126;46873,31889;46781,31653;46688,31417;46593,31181;46497,30946;46401,30712;46303,30477;46103,30011;46001,29778;45899,29546;45795,29314;45690,29083;45584,28853;45476,28623;45368,28394;45258,28165;45147,27936;45035,27709;44922,27481;44808,27255;44576,26803;44458,26579;44339,26354;44219,26131;44098,25908;43976,25685;43852,25463;43728,25242;43602,25022;43475,24802;43348,24583;43219,24364;43089,24146;42825,23712;42692,23496;42558,23281;42422,23066;42285,22852;42148,22639;42009,22426;41869,22215;41728,22004;41586,21793;41443,21583;41299,21374;41154,21166;41008,20959;40861,20752;40713,20546;40413,20136;40262,19932;40109,19730;39956,19527;39802,19326;39646,19125;39490,18925;39332,18726;39174,18528;38854,18134;38692,17938;38530,17743;38367,17549;38037,17163;37870,16971;37703,16781;37535,16590;37366,16401;37195,16213;37024,16026;36852,15839;36679,15653;36505,15469;36330,15285;36155,15102;35978,14919;35800,14738;35622,14558;35442,14378;35262,14200;35081,14022;34898,13845;34715,13670;34531,13495;34347,13321;34161,13148;33974,12976;33787,12805;33599,12634;33410,12465;33219,12297;33029,12130;32837,11963;32451,11633;32257,11470;32062,11308;31866,11146;31472,10826;31274,10668;31075,10510;30875,10354;30674,10198;30473,10044;30270,9891;30068,9738;29864,9587;29454,9287;29248,9139;29041,8992;28834,8846;28626,8701;28417,8557;28207,8414;27996,8272;27785,8131;27574,7991;27361,7852;27148,7715;26934,7578;26719,7442;26504,7308;26288,7175;25854,6911;25636,6781;25417,6652;25198,6525;24978,6398;24758,6272;24537,6148;24315,6024;24092,5902;23869,5781;23646,5661;23421,5542;23197,5424;22745,5192;22519,5078;22291,4965;22064,4853;21835,4742;21606,4632;21377,4524;21147,4416;20917,4310;20686,4205;20454,4101;20222,3999;19989,3897;19523,3697;19288,3599;19054,3503;18819,3407;18583,3312;18347,3219;18111,3127;17874,3036;17398,2858;17160,2771;16921,2685;16682,2600;16442,2516;16202,2434;15962,2353;15721,2273;15480,2194;15238,2116;14996,2040;14754,1965;14511,1891;14268,1818;13780,1676;13536,1607;13291,1540;13047,1473;12801,1408;12556,1344;12310,1281;12064,1220;11817,1159;11570,1100;11323,1042;11076,986;10828,930;10580,876;10332,824;10083,772;9834,722;9585,673;9336,625;9087,578;8837,533;8587,489;8337,446;8086,405;7836,365;7585,326;7334,288;7083,252;6831,217;6580,183;6328,150;6076,119;5824,89;5572,60;5320,32;5067,6;4815,-19;4309,-65;4056,-86;3803,-106;3297,-142;3044,-158;2790,-172;2537,-186;2283,-198;2030,-209;1776,-219;1523,-227;1269,-234;1015,-240;761,-244;508,-247;254,-249)`  

and when I try to use for-loop to get the points, VS code  says ,the polygon object is not iterable.

While the `for pt in polygon.to_dtype(gf.kcl.dbu)`, VS code raise an error, says 'DPolygon' object is not iterable. Could you tell me where can I read the documentation about these. I searched on the GDSFactory github website, but I did not find anything related.

**Reply by sebastian-goeldi** [reply]
Oh, I didn't see, the square brackets are at the wrong place: `[pt for polygon in dpoint_region.each() for pt in polygon.to_dtype(gf.kcl.dbu).each_point_hull()]` sorry. I guess I shouldn't just write the code from my head :smile:

**Reply by luomeng007** [reply]
Ok, now it works well. Thank you. After I understand this method and then I can apply this method to my own cases.

**Comment by sebastian-goeldi** [comment]
You can however force the optimization of DPolygon with `dpolygon.compress(True)` if you want to do that

**Reply by luomeng007** [reply]
Ok, I get it. Thanks. 

By this way, I would like to ask, is there some way to get the minimum distance between two random pcells in GDSFactory?

**Reply by sebastian-goeldi** [reply]
Yes, but it's expensive. I can give you an example later today (I'll edit this comment)./Don't expect miracles as it will be bad at runtime.

The easiest way to determine that you are **not** below a minimum threshold is way easier, and can easily be achieved with a min_width DRC.

**Reply by luomeng007** [reply]
Waiting for you...

**Reply by sebastian-goeldi** [reply]
The way that you would do this is create two regions (with `c.begin_shapes_rec(layer_of_interest_index)` as the constructor or even filtered down). Then create edge pairs from the two regions and use `min(ep.distance() for ep in region.hulls().edges().each())`. This is the short and sweet way I can come up with.

sorry for the delay
